azurerm/reader.go

```diff
@@ -0,0 +1,94 @@
+package azurerm
+
+import (
+	"context"
+	"fmt"
+
+	azureResourcesAPI "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/hashicorp/go-azure-helpers/authentication"
+	"github.com/hashicorp/go-azure-helpers/sender"
+)
+
+//go:generate go run ./cmd
+
+// AzureReader is the middleware between TC and AzureRM
+type AzureReader struct {
+	config     authentication.Config
+	authorizer autorest.Authorizer
+
+	resourceGroup azureResourcesAPI.Group
+}
+
+// NewAzureReader returns a AzureReader
+func NewAzureReader(ctx context.Context, clientID, clientSecret, environment, resourceGroupName, subscriptionID, tenantID string) (*AzureReader, error) {
+	// Config
+	cfgBuilder := &authentication.Builder{
+		ClientID:       clientID,
+		ClientSecret:   clientSecret,
+		Environment:    environment,
+		SubscriptionID: subscriptionID,
+		TenantID:       tenantID,
+		// ClientCertPassword: d.Get("client_certificate_password").(string),
+		// ClientCertPath:     d.Get("client_certificate_path").(string),
+
+		// SupportsAzureCliToken:    true,
+		SupportsClientSecretAuth: true,
+		// SupportsClientCertAuth:   true,
+	}
+
+	cfg, err := cfgBuilder.Build()
+	if err != nil {
+		return nil, fmt.Errorf("could not build 'azure/authentication.Config' because: %s", err)
+	}
+
+	// Authorizer
+	env, err := authentication.DetermineEnvironment(cfg.Environment)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize 'azure.Environment.' because: %s", err)
+	}
+
+	oauthConfig, err := cfg.BuildOAuthConfig(env.ActiveDirectoryEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize 'azure/authentication.OAuthConfig.' because: %s", err)
+	}
+	if oauthConfig == nil {
```

Comment:
`oauthConfig` can be nil even if there is no error ?

Reply:
I've taken this part from Azure SDK go examples, so it seems like it does.